src/pykeen/triples/statement_factory.py

```diff
@@ -0,0 +1,600 @@
+    # def num_entities(self) -> int:
+    #     """The number of unique entities."""
+    #     return len(self.entity_to_id)
+    #
+    # @property
+    # def num_relations(self) -> int:
+    #     """The number of unique relations."""
+    #     return len(self.relation_to_id)
+
+    @property
+    def statement_length(self) -> int:
+        """The number of unique relations."""
+        return self.max_num_qualifier_pairs * 2 + 3
+
+    @property
+    def num_statements(self) -> int:
+        """The number of statements."""
+        return self.mapped_statements.shape[0]
+
+    @property
+    def qualifier_ratio(self) -> float:
+        """Return the percentage of statements with qualifiers."""
+        return (~(self.mapped_statements[:, 3::2] == self.padding_idx).all(dim=1)).sum().item() / self.num_statements
+
+    def extra_repr(self) -> str:
+        """Extra representation."""
+        return (
+            f"num_entities={self.num_entities:,}, "
+            f"num_relations={self.num_relations:,}, "
+            f"num_statements={self.num_statements:,}, "
+            f"max_num_qualifier_pairs={self.max_num_qualifier_pairs}, "
+            f"qualifier_ratio={self.qualifier_ratio:.2%}"
+        )
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({self.extra_repr()})"
+
+    @property
+    def num_triples(self) -> int:
+        return self.num_statements
+
+
+def create_statement_entity_mapping(statements: Iterable[str]):
+    entities = list(sorted(set([e for s in statements for e in s[::2]])))  # padding is already in the ndarray
+    entity_to_id = {entity: id for id, entity in enumerate(entities)}
+    return entity_to_id
+
+
+def create_statement_relation_mapping(relations: np.ndarray):
+    # Sorting ensures consistent results when the triples are permuted
+    relation_labels = sorted(
+        set(relations.flatten()),
+        key=lambda x: (re.sub(f"{INVERSE_SUFFIX}$", "", x), x.endswith(f"{INVERSE_SUFFIX}")),
+    )
+    # Create mapping
+    return {str(label): i for (i, label) in enumerate(relation_labels)}
+
+
+def _create_multi_label_tails_instance(
+    mapped_statements: torch.Tensor,
+    use_tqdm: Optional[bool] = None,
+) -> Dict[Tuple[int, int], List[int]]:
+    """Create for each (h,r,q*) pair the multi tail label."""
+    logger.debug("Creating multi label tails instance")
+
+    """
+    The mapped triples matrix has to be a numpy array to ensure correct pair hashing, as explained in
+    https://github.com/pykeen/pykeen/commit/1bc71fe4eb2f24190425b0a4d0b9d6c7b9c4653a
+    """
+    mapped_statements = mapped_statements.cpu().detach().numpy()
+
+    s_p_q_to_multi_tails_new = _create_multi_label_instances(
+        mapped_statements,
+        element_1_index=0,
+        element_2_index=1,
+        label_index=2,
+        use_tqdm=use_tqdm,
+    )
+
+    logger.debug("Created multi label tails instance")
+
+    return s_p_q_to_multi_tails_new
+
+
+def _create_multi_label_instances(
```

Comment:
can we re-use this code for normal triples? If yes, then we can just extend the existing function.